torchaudio/csrc/ray_tracing.cpp

```diff
@@ -0,0 +1,577 @@
+/*
+Copyright (c) 2014-2017 EPFL-LCAV
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+/**
+ * Ray tracing implementation. This is heavily based on PyRoomAcoustics:
+ * https://github.com/LCAV/pyroomacoustics
+ */
+#include <torch/script.h>
+#include <torch/torch.h>
+#include <cmath>
+
+namespace torchaudio {
+namespace rir {
+namespace {
+
+const bool IS_HYBRID_SIM =
+    false; // TODO: remove this once hybrid method is supported
+const int ISM_ORDER = 10; // TODO: remove this once ISM method is supported
+#define EPS ((scalar_t)(1e-5))
+#define VAL(t) (*((t).template data_ptr<scalar_t>()))
+
+/**
+ * Wall helper class. A wall records its own absorption, reflection and
+ * scattering coefficient, and exposes a few methods for geometrical operations
+ * (e.g. reflection of a ray)
+ */
+template <typename scalar_t>
+class Wall {
+ public:
+  Wall(
+      const torch::Tensor& _absorption,
+      const torch::Tensor& _scattering,
+      const torch::Tensor& _normal,
+      const torch::Tensor& _origin)
+      : absorption(std::move(_absorption)),
+        reflection((scalar_t)1. - _absorption),
+        scattering(std::move(_scattering)),
+        normal(std::move(_normal)),
+        origin(std::move(_origin)) {}
```

Comment:
Applying `std::move` to `const` reference is strange.
You can probably remove `std::move` and copy construct them.